core/message-handling.go

```diff
@@ -251,28 +255,56 @@ func makeMessageStreamHandler(handle incomingMessageHandler, logger *logging.Log
 	}
 }
 
+var forwardOut map[uint32](chan messages.Request)
+
 // startPeerConnections initiates asynchronous message exchange with
 // peer replicas.
 func startPeerConnections(replicaID, n uint32, connector api.ReplicaConnector, log messagelog.MessageLog, logger *logging.Logger) error {
 	supply := makePeerMessageSupplier(log)
+	forwardOut = make(map[uint32](chan messages.Request))
 
 	for peerID := uint32(0); peerID < n; peerID++ {
 		if peerID == replicaID {
 			continue
 		}
 
+		forwardOut[peerID] = make(chan messages.Request)
+
 		connect := makePeerConnector(peerID, connector)
-		if err := startPeerConnection(connect, supply); err != nil {
+		forward := makeForwardedRequestSupplier(peerID)
+		if err := startPeerConnection(connect, supply, forward, forwardOut[peerID]); err != nil {
 			return fmt.Errorf("Cannot connect to replica %d: %s", peerID, err)
 		}
 	}
 
 	return nil
 }
 
+// makeForwardedRequestSupplier constructs a forwardedRequestSupplier using
+// the supplied replica ID.
+func makeForwardedRequestSupplier(peerID uint32) forwardedRequestSupplier {
+	return func(out chan<- []byte, in <-chan messages.Request) {
+		for {
+			msg := <-in
+			msgBytes, err := msg.MarshalBinary()
+			if err != nil {
+				panic(err)
+			}
+			out <- msgBytes
+		}
+	}
+}
+
+// forwardRequestMessage send Request message to primary replica when
+// prepare timer expires.
+func forwardRequestMessage(peerID uint32, msg messages.Request, logger *logging.Logger) {
+	logger.Debugf("Forward REQUEST to replica %d\n", peerID)
+	forwardOut[peerID] <- msg
```

Comment:
I'm afraid this may block the caller for indefinite time if the destination replica is faulty. How do we handle this situation?

Reply:
I didn't consider such situation. I think that this code does not block as long as `forwardOut` channel is listened by the goroutine running forwardedRequestSupplier, which is satisfied if the replica is not faulty. How does the destination replica affect this code?